editoast/src/views/timetable/track_occupancy.rs

```diff
@@ -175,69 +96,93 @@ pub fn find_track_occupancy_for_operational_point(
-        PathfindingResult::Success(pathfinding_result_success) => {
-            assert_eq!(
-                pathfinding_result_success.path_item_positions.len(),
-                train_schedule.path.len()
-            );
-            (
-                &pathfinding_result_success.path.track_section_ranges,
-                &pathfinding_result_success.path_item_positions,
-            )
-        }
-        PathfindingResult::Failure(_) => {
-            // The timetable can have a pathfinding failure for a train schedule
-            // because the train schedule is not valid.
-            // We skip it because we don't want to compute the track occupancy for an invalid train schedule.
-            // It's not a problem because we are just looking for the track occupancy of the operational point for the valid trains.
-            tracing::info!(train_schedule.train_name, "pathfinding failed");
-            return vec![];
+    let (is_valid, final_output, pathfinding_success) = match (simulation, pathfinding) {
+        (
+            simulation::Response::Success(SimulationResponseSuccess { final_output, .. }),
+            PathfindingResult::Success(pathfinding_success),
+        ) => (true, Some(final_output), Some(pathfinding_success)),
+        (_, PathfindingResult::Success(pathfinding_success)) => {
+            (false, None, Some(pathfinding_success))
         }
+        _ => (false, None, None),
     };
-    let path_projection = PathProjection::new(track_ranges);
 
-    // Get the positions and the times from the simulation
-    let report_train = match simulation {
-        simulation::Response::Success(SimulationResponseSuccess { final_output, .. }) => {
-            &final_output.report_train
-        }
-        _ => {
-            tracing::info!(train_schedule.train_name, "simulation failed");
+    if let Some(matching_index) =
+        find_matching_path_item_index(&train_schedule.path, operational_point_id, path_item_cache)
+    {
+        let schedule_item = train_schedule
+            .schedule
+            .iter()
+            .find(|schedule| schedule.at.0 == train_schedule.path[matching_index].id.0);
+
+        // For invalid trains, skip if this operational point has no scheduled arrival
+        // (except for the first path item which arrival_time is t=0)
+        if !is_valid && matching_index != 0 && schedule_item.is_none_or(|s| s.arrival.is_none()) {
             return vec![];
         }
-    };
 
-    // Find matching path item
-    let path_item_id = match_path_item_id_with_operational_point(
-        path_item_cache,
-        train_schedule,
-        operational_point_id,
-    );
+        let arrival_time = final_output
+            .map(|output| output.report_train.path_item_times[matching_index] as i64)
+            .or_else(|| {
+                schedule_item
+                    .and_then(|s| s.arrival.as_ref())
+                    .map(|a| a.num_milliseconds())
+            })
+            .unwrap_or(0);
```

Comment:
We can handle the first case, retrieving the `arrival_time`. Actually, the `unwrap_or(0)` is "useless" since this case will never happen.


```suggestion
        let Some(arrival_time) = final_output
            .map(|output| output.report_train.path_item_times[matching_index] as i64)
            .or_else(|| {
                schedule_item
                    .and_then(|s| s.arrival.as_ref())
                    .map(|a| a.num_milliseconds())
            })
        else {
            // For invalid trains, skip if this operational point has no scheduled arrival
            // (except for the first path item which arrival_time is t=0)
            return vec![];
        };
```

Reply:
The arrival_time for the first path_item should always be 0, because it represents the train's starting point. In this specific case, there is no arrival value in the corresponding schedule_item, so I think we need a special check to handle this: if the matching path_item is the first in the path (index 0), we must set arrival_time to 0 even if there is no arrival in the schedule.

Reply:
I simplified the workflow but kept the specific case for the first operational point.